pkg/lib/synthesize/synthesis.go

```diff
@@ -0,0 +1,679 @@
+		return err
+	}
+
+	s.synthesizedCSV = &v1alpha1.ClusterServiceVersion{
+		TypeMeta: v1.TypeMeta{
+			APIVersion: v1alpha1.ClusterServiceVersionAPIVersion,
+			Kind:       CSVKind,
+		},
+		ObjectMeta: v1.ObjectMeta{
+			// TODO: Are there other fields can be set here?
+			Name:         csvBasics.Name,
+			GenerateName: csvBasics.GenerateName,
+		},
+		Spec: v1alpha1.ClusterServiceVersionSpec{
+			InstallStrategy:           *installStrategy,
+			CustomResourceDefinitions: *specCRDs,
+			APIServiceDefinitions:     *apiServiceDefs,
+			Version:                   csvBasics.Version,
+			MinKubeVersion:            csvBasics.MinKubeVersion,
+			Maturity:                  csvBasics.Maturity,
+			DisplayName:               csvBasics.DisplayName,
+			Description:               csvBasics.Description,
+			Keywords:                  csvBasics.Keywords,
+			Maintainers:               csvBasics.Maintainers,
+			Provider:                  csvBasics.Provider,
+			Links:                     csvBasics.Links,
+			Icon:                      csvBasics.Icon,
+			InstallModes:              csvBasics.InstallModes,
+			Replaces:                  csvBasics.Replaces,
+			Labels:                    csvBasics.Labels,
+			Annotations:               csvBasics.Annotations,
+			Selector:                  csvBasics.Selector,
+		},
+	}
+	return nil
+}
+
+// loadStructuredObjects gets supported structured object recursively from bundle directory and put them into the map.
+// Map of the same bundle should always produce the same map considering directory walk is lexical ordered and therefore
+// input order is deterministic.
+// All supported type of structures are presented in supportedResources, extend the support types by adding to the map.
+func (s *synthesize) loadStructuredObjects() error {
+	return filepath.Walk(s.bundleDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		// Recursively walking through every file in the bundle directory.
+		if !info.IsDir() {
+			reader, err := os.Open(path)
+			if err != nil {
+				return err
+			}
+
+			bufObj := unstructured.Unstructured{}
+			decoder := yaml.NewYAMLOrJSONDecoder(reader, 30)
+			if err = decoder.Decode(&bufObj); err != nil {
+				// try to parse any file that does not have a kind in here.
+				if err := s.loadDependencies(path); err == nil {
+					return nil
+				}
+				s.logger.Warnf("file %s is not decodable as YAML or JSON, %v", info.Name(), err)
+				return nil
+			}
+
+			// accepting supported types and JSON or YAML files with no kind field.
+			if !IsSupported(bufObj.GetKind()) {
+				s.logger.Warnf("file %s with kind %s is not supported", info.Name(), bufObj.GetKind())
+			}
+
+			switch bufObj.GetKind() {
+			case CSVKind:
+				if s.existingCSVPath != "" {
+					return fmt.Errorf("more than one ClusterServiceVersion is found in bundle")
+				}
+				s.existingCSVPath = path
+				s.appendTostructuredObjectMap(CSVKind, bufObj)
+
+			case RegistryV2CSVKind:
```

Comment:
To follow the existing conventions, it may be worthwhile to recursively descend the working directory to find and use the `annotations.yaml` file’s directory (`metadata`) as the location for this new metadata file.

Reply:
Having CSV-metadata in a non-kube object in `/manifests` breaks our assumptions that everything in `/manifests` is kubectl-able. We should move this into the /metadata folder (more precisely - the folder containing annotations.json).